plugins/obs-filters/cross-stitching-filter.c

```diff
@@ -0,0 +1,259 @@
+ * OBS. Jim has added several useful functions to help keep memory leaks to
+ * a minimum, and handle the destruction and construction of these filters.
+ */
+static void cross_stitching_filter_destroy(void *data)
+{
+	struct cross_stitching_filter_data *filter = data;
+
+	if (filter->effect) {
+		obs_enter_graphics();
+		gs_effect_destroy(filter->effect);
+		obs_leave_graphics();
+	}
+
+	bfree(data);
+}
+
+/*
+ * When you apply a filter OBS creates it, and adds it to the source. OBS also
+ * starts rendering it immediately. This function doesn't just 'create' the
+ * filter, it also calls the render function (farther below) that contains the
+ * actual rendering code.
+ */
+static void *cross_stitching_filter_create(obs_data_t *settings, obs_source_t *context)
+{
+	/*
+	 * Because of limitations of pre-C99 compilers, you can't create an
+	 * array that doesn't have a know size at compile time. The below
+	 * function calculates the size needed and allocates memory to
+	 * handle the source.
+	 */
+	struct cross_stitching_filter_data *filter =
+		bzalloc(sizeof(struct cross_stitching_filter_data));
+
+	/*
+	 * By default the effect file is stored in the ./data directory that
+	 * your filter resides in.
+	 */
+	#ifdef WIN32
+		char *effect_path = obs_module_file("cross_stitching_filter.effect");
+	#else
+		char *effect_path = obs_module_file("cross_stitching_filter_gl.effect");
+	#endif
+
+	filter->context = context;
+
+	/* Here we enter the GPU drawing/shader portion of our code. */
+	obs_enter_graphics();
+
+	/* Load the shader on the GPU. */
+	filter->effect = gs_effect_create_from_file(effect_path, NULL);
+
+	/* If the filter is active pass the parameters to the filter. */
+	if (filter->effect) {
+		filter->stitch_gap_param = gs_effect_get_param_by_name(
+			filter->effect, SETTING_STITCH_GAP);
+		filter->negative_bias_param = gs_effect_get_param_by_name(
+			filter->effect, SETTING_NEGATIVE_BIAS);
+		filter->stitch_color_param = gs_effect_get_param_by_name(
+			filter->effect, SETTING_STITCH_COLOR);
+		filter->invert_param = gs_effect_get_param_by_name(
+			filter->effect, SETTING_INVERT);
+		filter->greyscale_param = gs_effect_get_param_by_name(
+			filter->effect, SETTING_GREYSCALE);
+	}
+
+	obs_leave_graphics();
+
+	bfree(effect_path);
+
+	if (!filter->effect) {
+		cross_stitching_filter_destroy(filter);
+		return NULL;
+	}
+
+	/*
+	 * It's important to call the update function here. if we don't
+	 * we could end up with the user controlled sliders and values
+	 * updating, but the visuals not updating to match.
+	 */
+	cross_stitching_filter_update(filter, settings);
+	return filter;
+}
+
+/* This is where the actual rendering of the filter takes place. */
+static void cross_stitching_filter_render(void *data, gs_effect_t *effect)
+{
+	struct cross_stitching_filter_data *filter = data;
+
+	if (!obs_source_process_filter_begin(filter->context, GS_RGBA,
+				OBS_ALLOW_DIRECT_RENDERING))
```

Comment:
This is indented with three tabs.  I've seen both one tab, two tabs, and spaces used for indenting continued `if` statement conditions.  I do not recall which is correct here.  Jim, of course, is the final arbiter there.